SimulationAgent/SimulationManager.cs

```diff
@@ -0,0 +1,500 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using System.Linq;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Clustering;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.DataStructures;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Diagnostics;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Exceptions;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Models;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Runtime;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.SimulationAgent.DeviceConnection;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.SimulationAgent.DeviceProperties;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.SimulationAgent.DeviceState;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.SimulationAgent.DeviceTelemetry;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.SimulationAgent
+{
+    public interface ISimulationManager
+    {
+        Task InitAsync(
+            Simulation simulation,
+            ConcurrentDictionary<string, IDeviceStateActor> deviceStateActors,
+            ConcurrentDictionary<string, IDeviceConnectionActor> deviceConnectionActors,
+            ConcurrentDictionary<string, IDeviceTelemetryActor> deviceTelemetryActors,
+            ConcurrentDictionary<string, IDevicePropertiesActor> devicePropertiesActors);
+
+        // === BEGIN - Executed by Agent.RunAsync
+
+        // Ensure the partitions assigned to this node remain assigned
+        Task HoldAssignedPartitionsAsync();
+
+        // Check if there are new partitions
+        Task AssignNewPartitionsAsync();
+
+        // Check if there are partitions changes and act accordingly
+        Task HandleAssignedPartitionChangesAsync();
+
+        // Check if the cluster size has changed and act accordingly
+        Task UpdateThrottlingLimitsAsync();
+
+        void PrintStats();
+
+        // === END - Executed by Agent.RunAsync
+
+        // Stop all the actors and delete them - executed by Agent.StopInactiveSimulations
+        void TearDown();
+
+        // // Executed by DeviceConnectionTask.RunAsync
```

Comment:
Remove

Reply:
I want to leave this code in for now. It's an optimization that I'll address in a subsequent PR.